### PR TITLE
Added a CppComponentDetector

### LINKED ISSUE #64: Create a component detector for C++ (~60 new lines of code)
We'd like to measure the quality of apps written in C++.

There are "component detectors" to determine frameworks, platform, and type of an application.

It's necessary to implement **CppComponentDetector** the similar way as is [JavaComponentDetector](https://github.com/DXHeroes/dx-scanner/pull/50/files#diff-713f3c9ac097761276975cddf91b94b1) implemented. Because C++ is not a front-end language, so it's not necessary to recognize Frontend/Backend platform. 
It's around 60 new lines of code including tests.

More info in [our Contribution guide](https://github.com/DXHeroes/dx-scanner/blob/master/CONTRIBUTING.md).

Other sources:
- [JavaComponentDetector](https://github.com/DXHeroes/dx-scanner/pull/50/files#diff-713f3c9ac097761276975cddf91b94b1)
- [JavaScriptComponentDetector (including FE&BE platform recognition)](https://github.com/DXHeroes/dx-scanner/blob/master/src/detectors/JavaScript/JavaScriptComponentDetector.ts)

### PR DESCRIPTION
**Pull request by iamRishabh07**
## Description

Fixes #64 
Added a C++ Component detector which will measure the quality of apps written in C++.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated the documentation accordingly.
- [x] I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by adelkahomolova** [comment]
Thank you for your PR! @iamRishabh07 Can you add tests for your code? Thanks :)

**Comment by adelkahomolova** [comment]
Also, you need to add code to [languageContextBinding.ts](https://github.com/DXHeroes/dx-scanner/blob/master/src/contexts/language/languageContextBinding.ts) and [index.ts](https://github.com/DXHeroes/dx-scanner/blob/master/src/detectors/index.ts). Your code wouldn't be working without these additions. @iamRishabh07

**Comment by iamRishabh07** [comment]
@adelkahomolova Can you provide me any reference on what kind of test to add.

**Comment by adelkahomolova** [comment]
Look at [JavaScriptComponentDetector.spec.ts](https://github.com/DXHeroes/dx-scanner/blob/master/src/detectors/JavaScript/JavaScriptComponentDetector.spec.ts), [JavaComponentDetector.spec.ts](https://github.com/DXHeroes/dx-scanner/blob/master/src/detectors/Java/JavaComponentDetector.spec.ts) or any other tests for component detectors in [detectors folder](https://github.com/DXHeroes/dx-scanner/tree/master/src/detectors) 😉

**Comment by adelkahomolova** [comment]
@iamRishabh07 Hi, how are you going with the tests and the small additions I mentioned above? :)

**Comment by iamRishabh07** [comment]
@adelkahomolova I have made the required changes. Do I also need to add the PackageInspector?

**Comment by adelkahomolova** [comment]
That would be awesome, if you'll add also Package Inspector. Because after that, the dxs-scanner will support C++. Thank you for the question. You can get the inspiration from [JavaScriptPackageInspector](https://github.com/DXHeroes/dx-scanner/blob/master/src/inspectors/package/JavaScriptPackageInspector.ts), or [JavaScriptPackageInspector](https://github.com/DXHeroes/dx-scanner/blob/master/src/inspectors/package/JavaScriptPackageInspector.ts) or any other package inspector from [package folder](https://github.com/DXHeroes/dx-scanner/tree/master/src/inspectors/package). Don't forget about tests. Thank you, @iamRishabh07  :)

**Comment by iamRishabh07** [comment]
@adelkahomolova what is the use of Gradle and Maven in JavaPackageInspector?

**Comment by adelkahomolova** [comment]
Both are similar to `package.json` for javascript. It's a build and dependency resolution tool for Java. @iamRishabh07

**Comment by adelkahomolova** [comment]
Hi @iamRishabh07 is there any progress? :)

**Comment by iamRishabh07** [comment]
@adelkahomolova I have made the changes as required. And for the package Inspector, I am still figuring it out.

File: src/contexts/language/LanguageContext.ts [hunk 1]
```diff
@@ -48,4 +48,4 @@ export class LanguageContext extends ContextBase {
   getProjectComponentContext(component: ProjectComponent): ProjectComponentContext {
     return this.projecComponentContextFactory(component);
   }
-}
+}
```
Comment (adelkahomolova) [inline src/contexts/language/LanguageContext.ts:51]: add a new line at end of a file
```suggestion
}

```

File: src/contexts/language/languageContextBinding.ts [hunk 2]
```diff
@@ -18,7 +19,6 @@ import { PackageInspectorBase } from '../../inspectors/package/PackageInspectorB
 import { IProjectComponentDetector } from '../../detectors/IProjectComponentDetector';
 import { ScanningStrategy } from '../../detectors';
 import { ProjectFilesBrowserService } from '../../services';
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:21]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 3]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
     },
   );
 };
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:32]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 4]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
 const createLanguageContainer = (languageAtPath: LanguageAtPath, rootContainer: Container): Container => {
   const container = rootContainer.createChild();
   container.bind(Types.LanguageAtPath).toConstantValue(languageAtPath);
   const scanningStrategy = container.get<ScanningStrategy>(Types.ScanningStrategy);
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:37]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 5]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
   const projectFilesBrowserService = container.get<ProjectFilesBrowserService>(Types.IProjectFilesBrowser);
   const fileInspector = container.get<FileInspector>(Types.IFileInspector);
   const fileInspectorForRoot = new FileInspector(projectFilesBrowserService, scanningStrategy.rootPath || fileInspector.basePath);
   container.bind(Types.IRootFileInspector).toConstantValue(fileInspectorForRoot);
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:42]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 6]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
   bindFileAccess(languageAtPath, container);
   bindComponentDetectors(container);
   bindProjectComponentContext(container);
   bindPackageInspectors(languageAtPath, container);
   bindCollaborationInspectors(container);
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:48]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 7]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
   container.bind(LanguageContext).toSelf();
   return container;
 };
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:52]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 8]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
 const bindFileAccess = (languageAtPath: LanguageAtPath, container: Container) => {
   container.bind(Types.FileInspectorBasePath).toConstantValue(languageAtPath.path);
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:55]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 9]
```diff
@@ -29,33 +29,26 @@ export const bindLanguageContext = (container: Container) => {
...
   container.bind(Types.IFileInspector).to(FileInspector).inSingletonScope();
 };
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:58]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 10]
```diff
@@ -79,19 +75,17 @@ const bindComponentDetectors = (container: Container) => {
       .whenTargetTagged(DETECT_LANGUAGE_TAG, current.value.detectedLanguage);
     current = iterator.next();
   }
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:82]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 11]
```diff
@@ -79,19 +75,17 @@ const bindComponentDetectors = (container: Container) => {
...
   container.bind(Types.ProjectComponentDetectorFactory).toFactory((ctx) => {
     return getProjectComponentDetectorFactory(ctx.container as Container);
   });
 };
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:87]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 12]
```diff
@@ -79,19 +75,17 @@ const bindComponentDetectors = (container: Container) => {
...
 const bindCollaborationInspectors = (container: Container) => {
   container.bind(Types.ICollaborationInspector).to(CollaborationInspector).inSingletonScope();
   container.bind(Types.IIssueTrackingInspector).to(IssueTrackingInspector).inSingletonScope();
 };
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-const resolveBindingPackageInspector = (packageInspector: { new (...args: any[]): PackageInspectorBase }, container: Container) => {
+const resolveBindingPackageInspector = (packageInspector: { new(...args: any[]): PackageInspectorBase }, container: Container) => {
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:88]:
```suggestion
const resolveBindingPackageInspector = (packageInspector: { new (...args: any[]): PackageInspectorBase }, container: Container) => {
```

File: src/contexts/language/languageContextBinding.ts [hunk 13]
```diff
@@ -104,14 +98,15 @@ const resolveBindingPackageInspector = (packageInspector: { new (...args: any[])
 
 const componentGenerator = function* (): Generator<{
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  componentDetector: { new (...args: any[]): IProjectComponentDetector };
+  componentDetector: { new(...args: any[]): IProjectComponentDetector };
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:101]:
```suggestion
  componentDetector: { new (...args: any[]): IProjectComponentDetector };
```

File: src/contexts/language/languageContextBinding.ts [hunk 14]
```diff
@@ -124,10 +119,9 @@ export const DETECT_LANGUAGE_TAG = 'language';
 export const detectsLanguage = (language: ProgrammingLanguage) => {
   return tagged(DETECT_LANGUAGE_TAG, language);
 };
-
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:127]: Why did you remove the line?

File: src/contexts/language/languageContextBinding.ts [hunk 15]
```diff
@@ -124,10 +119,9 @@ export const DETECT_LANGUAGE_TAG = 'language';
...
 export const getProjectComponentDetectorFactory = (container: Container) => {
   return (language: ProgrammingLanguage) => {
     const detectors = container.getAllTagged(Types.IProjectComponentDetector, DETECT_LANGUAGE_TAG, language);
     return detectors;
   };
-};
+};
```
Comment (adelkahomolova) [inline src/contexts/language/languageContextBinding.ts:127]: add new line
```suggestion
};

```

File: src/detectors/Cpp/CPPComponentDetector.spec.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import { CPPComponentDetector } from './CPPComponentDetector';
+import { ProgrammingLanguage, ProjectComponentPlatform } from '../../model';
+jest.mock('../../inspectors/package/CPPPackageInspector');
+
+describe('CPPComponentDetector', () => {
+    let detector: CPPComponentDetector;
+    const MockedCPPPackageInspector = <jest.Mock<CPPPackageInspector>>(<unknown>CPPPackageInpector);
```
Comment (adelkahomolova) [inline src/detectors/Cpp/CPPComponentDetector.spec.ts:7]:
```suggestion
    const MockedCPPPackageInspector = <jest.Mock<CPPPackageInspector>>(<unknown>CPPPackageInspector);
```

File: src/detectors/index.ts [hunk 17]
```diff
@@ -8,4 +10,4 @@ export * from './PHP/PHPLanguageDetector';
 export * from './Python/PythonLanguageDetector';
 export * from './Ruby/RubyLanguageDetector';
 export * from './ScanningStrategyDetector';
-export * from './IScanningStrategy';
+export * from './IScanningStrategy';
```
Comment (adelkahomolova) [inline src/detectors/index.ts:13]: add new line
```suggestion
export * from './IScanningStrategy';

```

File: src/model.ts [hunk 18]
```diff
@@ -198,4 +198,4 @@ export enum GitFLow {
   Git = 'Git',
 }
 
-export type VCSService = GitHubService | BitbucketService | GitLabService;
+export type VCSService = GitHubService | BitbucketService | GitLabService;
```
Comment (adelkahomolova) [inline src/model.ts:201]: add new line
```suggestion
export type VCSService = GitHubService | BitbucketService | GitLabService;

```

**Review by adelkahomolova — CHANGES_REQUESTED** [review]

**Comment by adelkahomolova** [comment]
Hi, @iamRishabh07 , is there any progress? :)

**Pull request closed without merging**